From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Prathamesh Shete <pshete@nvidia.com>
Date: Wed, 21 May 2025 05:03:17 +0000
Subject: [PATCH 130/243] NVIDIA: SAUCE: pinctrl: tegra: Enhance suspend/resume
 to save per-pingroup configuration

Previously, the Tegra pinctrl suspend/resume logic saved and restored
registers sequentially per bank, without considering individual pingroup
layout or relevance. This approach could result in saving unnecessary
registers and made the backup format fragile to changes in bank size or
layout.

This patch enhances the suspend and resume logic to explicitly save and
restore only the relevant registers associated with each pingroup,
using the mux/pupd/tri/drive and associated bank/offset information
from the SoC-specific pingroup definitions.

Key improvements:
- Save and restore only meaningful registers per pingroup.
- Avoid assumptions about contiguous memory layout of pinctrl banks.

Bug 5250926
Bug 5143785

Signed-off-by: Prathamesh Shete <pshete@nvidia.com>
Change-Id: I8ca7d58c2e2916510fb6aa179ba19d5a24a47a59
(cherry picked from commit e2eca35319600c37be0561eaab2b4100165cb641)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3503641
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
(cherry picked from commit 17c772f4d0c1d3d466b6a25ff11f3c3fc6d6a5ec)

diff --git a/drivers/pinctrl/tegra/pinctrl-tegra.c b/drivers/pinctrl/tegra/pinctrl-tegra.c
index 0cb7a705a5243..2282a96af025c 100644
--- a/drivers/pinctrl/tegra/pinctrl-tegra.c
+++ b/drivers/pinctrl/tegra/pinctrl-tegra.c
@@ -23,6 +23,7 @@
 #include <linux/pinctrl/pinconf.h>
 #include <linux/pinctrl/pinctrl.h>
 #include <linux/pinctrl/pinmux.h>
+#include <soc/tegra/virt/hv-ivc.h>
 
 #include "../core.h"
 #include "../pinctrl-utils.h"
@@ -754,30 +755,46 @@ static void tegra_pinctrl_clear_parked_bits(struct tegra_pmx *pmx)
 	}
 }
 
-static size_t tegra_pinctrl_get_bank_size(struct device *dev,
-					  unsigned int bank_id)
-{
-	struct platform_device *pdev = to_platform_device(dev);
-	struct resource *res;
-
-	res = platform_get_resource(pdev, IORESOURCE_MEM, bank_id);
-
-	return resource_size(res) / 4;
-}
-
 static int tegra_pinctrl_suspend(struct device *dev)
 {
 	struct tegra_pmx *pmx = dev_get_drvdata(dev);
 	u32 *backup_regs = pmx->backup_regs;
-	u32 __iomem *regs;
-	size_t bank_size;
-	unsigned int i, k;
+	unsigned int i;
+	const struct tegra_pingroup *g;
+	u32 val;
 
-	for (i = 0; i < pmx->nbanks; i++) {
-		bank_size = tegra_pinctrl_get_bank_size(dev, i);
-		regs = pmx->regs[i];
-		for (k = 0; k < bank_size; k++)
-			*backup_regs++ = readl_relaxed(regs++);
+#if IS_ENABLED(CONFIG_TEGRA_HV_DRIVER)
+	if (is_tegra_hypervisor_mode())
+		return 0;
+#endif
+
+	/* Save all register values for each pingroup */
+	for (i = 0; i < pmx->soc->ngroups; i++) {
+		g = &pmx->soc->groups[i];
+
+		/* Save mux register which includes all pin configurations */
+		if (g->mux_reg >= 0) {
+			val = pmx_readl(pmx, g->mux_bank, g->mux_reg);
+			*backup_regs++ = val;
+		}
+
+		/* Save pull-up/pull-down register */
+		if (g->pupd_reg >= 0) {
+			val = pmx_readl(pmx, g->pupd_bank, g->pupd_reg);
+			*backup_regs++ = val;
+		}
+
+		/* Save drive strength register */
+		if (g->drv_reg >= 0) {
+			val = pmx_readl(pmx, g->drv_bank, g->drv_reg);
+			*backup_regs++ = val;
+		}
+
+		/* Save tristate register */
+		if (g->tri_reg >= 0) {
+			val = pmx_readl(pmx, g->tri_bank, g->tri_reg);
+			*backup_regs++ = val;
+		}
 	}
 
 	return pinctrl_force_sleep(pmx->pctl);
@@ -787,21 +804,50 @@ static int tegra_pinctrl_resume(struct device *dev)
 {
 	struct tegra_pmx *pmx = dev_get_drvdata(dev);
 	u32 *backup_regs = pmx->backup_regs;
-	u32 __iomem *regs;
-	size_t bank_size;
-	unsigned int i, k;
+	unsigned int i;
+	const struct tegra_pingroup *g;
+	u32 val;
 
-	for (i = 0; i < pmx->nbanks; i++) {
-		bank_size = tegra_pinctrl_get_bank_size(dev, i);
-		regs = pmx->regs[i];
-		for (k = 0; k < bank_size; k++)
-			writel_relaxed(*backup_regs++, regs++);
+#if IS_ENABLED(CONFIG_TEGRA_HV_DRIVER)
+	if (is_tegra_hypervisor_mode())
+		return 0;
+#endif
+
+	/* Restore all register values for each pingroup */
+	for (i = 0; i < pmx->soc->ngroups; i++) {
+		g = &pmx->soc->groups[i];
+
+		/* Restore mux register which includes all pin configurations */
+		if (g->mux_reg >= 0) {
+			val = *backup_regs++;
+			pmx_writel(pmx, val, g->mux_bank, g->mux_reg);
+		}
+
+		/* Restore pull-up/pull-down register */
+		if (g->pupd_reg >= 0) {
+			val = *backup_regs++;
+			pmx_writel(pmx, val, g->pupd_bank, g->pupd_reg);
+		}
+
+		/* Restore drive strength register */
+		if (g->drv_reg >= 0) {
+			val = *backup_regs++;
+			pmx_writel(pmx, val, g->drv_bank, g->drv_reg);
+		}
+
+		/* Restore tristate register */
+		if (g->tri_reg >= 0) {
+			val = *backup_regs++;
+			pmx_writel(pmx, val, g->tri_bank, g->tri_reg);
+		}
 	}
 
-	/* flush all the prior writes */
-	readl_relaxed(pmx->regs[0]);
-	/* wait for pinctrl register read to complete */
-	rmb();
+	/* Clear any parked bits that might have been set during suspend */
+	tegra_pinctrl_clear_parked_bits(pmx);
+
+	/* Ensure all writes are completed */
+	readl(pmx->regs[0]);
+
 	return 0;
 }
 
@@ -846,6 +892,25 @@ int tegra_pinctrl_probe(struct platform_device *pdev,
 	if (!pmx->pingroup_configs)
 		return -ENOMEM;
 
+	/* Calculate total number of registers we need to save */
+	for (i = 0; i < pmx->soc->ngroups; i++) {
+		const struct tegra_pingroup *g = &pmx->soc->groups[i];
+
+		if (g->mux_reg >= 0)
+			backup_regs_size += sizeof(u32);
+		if (g->pupd_reg >= 0)
+			backup_regs_size += sizeof(u32);
+		if (g->drv_reg >= 0)
+			backup_regs_size += sizeof(u32);
+		if (g->tri_reg >= 0)
+			backup_regs_size += sizeof(u32);
+	}
+
+	/* Allocate backup memory based on actual number of registers needed */
+	pmx->backup_regs = devm_kzalloc(&pdev->dev, backup_regs_size, GFP_KERNEL);
+	if (!pmx->backup_regs)
+		return -ENOMEM;
+
 	/*
 	 * Each mux group will appear in 4 functions' list of groups.
 	 * This over-allocates slightly, since not all groups are mux groups.
@@ -904,7 +969,6 @@ int tegra_pinctrl_probe(struct platform_device *pdev,
 		res = platform_get_resource(pdev, IORESOURCE_MEM, i);
 		if (!res)
 			break;
-		backup_regs_size += resource_size(res);
 	}
 	pmx->nbanks = i;
 
@@ -913,11 +977,6 @@ int tegra_pinctrl_probe(struct platform_device *pdev,
 	if (!pmx->regs)
 		return -ENOMEM;
 
-	pmx->backup_regs = devm_kzalloc(&pdev->dev, backup_regs_size,
-					GFP_KERNEL);
-	if (!pmx->backup_regs)
-		return -ENOMEM;
-
 	for (i = 0; i < pmx->nbanks; i++) {
 		pmx->regs[i] = devm_platform_ioremap_resource(pdev, i);
 		if (IS_ERR(pmx->regs[i]))
-- 
2.43.0

